torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -109,52 +110,125 @@ std::shared_ptr<FutureMessage> sendPythonRemoteCall(
 
 using namespace torch::distributed::autograd;
 
-std::shared_ptr<FutureIValue> toFutureIValue(
-    const std::shared_ptr<FutureMessage>& fm,
+c10::intrusive_ptr<JitFuture> wrapFutureMessageInJitFuture(
+    const std::shared_ptr<FutureMessage>& futureResponseMessage,
     bool hasValue) {
-  auto fv = std::make_shared<FutureIValue>();
-
-  fm->addCallback([fv, hasValue](const FutureMessage& fm) {
-    if (fm.hasError()) {
-      fv->setError(*fm.error());
-    } else if (hasValue) {
-      // Don't need to acquire GIL here, as toPyObj and toIValue acquires GIL
-      // when creating/copying the py::object
-      fv->markCompleted(
-          jit::toIValue(toPyObj(fm.constValue()), PyObjectType::get()));
-    } else {
-      fv->markCompleted(IValue());
-    }
-  });
-
-  return fv;
+  if (hasValue) {
+    // NB: The custom deleter is necessary because the FutureIValue object
+    // holds a py::object and it would require GIL to delete.
+    c10::intrusive_ptr<JitFuture> jitFuture =
+        c10::make_intrusive<JitFuture>(PyObjectType::get());
+
+    futureResponseMessage->addCallback(
+        [jitFuture](const FutureMessage& futureResponseMessage) {
+          // Don't need to acquire GIL here, as toPyObj acquires GIL
+          // when creating the py::object
+          if (futureResponseMessage.hasError()) {
+            jitFuture->setError(futureResponseMessage.error()->what());
+          } else {
+            jitFuture->markCompleted(jit::toIValue(
+                toPyObj(futureResponseMessage.constValue()),
+                PyObjectType::get()));
+          }
+        });
+
+    return jitFuture;
+  } else {
+    c10::intrusive_ptr<JitFuture> jitFuture =
```

Comment:
Have to split them again because
ivalue::Future has a type field, which I believe is not necessary.
The 2 possibilities of futures here have different types. one will put in a python object. The other doesn't.
We can remove this branching after ivalue::Future is removed.